In [4]:
import pandas as pd




In [5]:
df = pd.read_csv("D:\Projects\healthcare-noshow-analysis\data\KaggleV2-May-2016.csv")


In [6]:
df

,PatientId,AppointmentID,Gender,ScheduledDay,AppointmentDay,Age,Neighbourhood,Scholarship,Hipertension,Diabetes,Alcoholism,Handcap,SMS_received,No-show
0,2.987250e+13,5642903,F,2016-04-29T18:38:08Z,2016-04-29T00:00:00Z,62,JARDIM DA PENHA,0,1,0,0,0,0,No
1,5.589978e+14,5642503,M,2016-04-29T16:08:27Z,2016-04-29T00:00:00Z,56,JARDIM DA PENHA,0,0,0,0,0,0,No
2,4.262962e+12,5642549,F,2016-04-29T16:19:04Z,2016-04-29T00:00:00Z,62,MATA DA PRAIA,0,0,0,0,0,0,No
3,8.679512e+11,5642828,F,2016-04-29T17:29:31Z,2016-04-29T00:00:00Z,8,PONTAL DE CAMBURI,0,0,0,0,0,0,No
4,8.841186e+12,5642494,F,2016-04-29T16:07:23Z,2016-04-29T00:00:00Z,56,JARDIM DA PENHA,0,1,1,0,0,0,No
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
110522,2.572134e+12,5651768,F,2016-05-03T09:15:35Z,2016-06-07T00:00:00Z,56,MARIA ORTIZ,0,0,0,0,0,1,No
110523,3.596266e+12,5650093,F,2016-05-03T07:27:33Z,2016-06-07T00:00:00Z,51,MARIA ORTIZ,0,0,0,0,0,1,No
110524,1.557663e+13,5630692,F,2016-04-27T16:03:52Z,2016-06-07T00:00:00Z,21,MARIA ORTIZ,0,0,0,0,0,1,No
110525,9.213493e+13,5630323,F,2016-04-27T15:09:23Z,2016-06-07T00:00:00Z,38,MARIA ORTIZ,0,0,0,0,0,1,No


In [7]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110527 entries, 0 to 110526
Data columns (total 14 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   PatientId       110527 non-null  float64
 1   AppointmentID   110527 non-null  int64  
 2   Gender          110527 non-null  object 
 3   ScheduledDay    110527 non-null  object 
 4   AppointmentDay  110527 non-null  object 
 5   Age             110527 non-null  int64  
 6   Neighbourhood   110527 non-null  object 
 7   Scholarship     110527 non-null  int64  
 8   Hipertension    110527 non-null  int64  
 9   Diabetes        110527 non-null  int64  
 10  Alcoholism      110527 non-null  int64  
 11  Handcap         110527 non-null  int64  
 12  SMS_received    110527 non-null  int64  
 13  No-show         110527 non-null  object 
dtypes: float64(1), int64(8), object(5)
memory usage: 11.8+ MB


In [8]:
df['ScheduledDay'] = pd.to_datetime(df['ScheduledDay'])
df['AppointmentDay'] = pd.to_datetime(df['AppointmentDay'])

In [9]:
print("Shape:", df.shape)

print("\nData Types:\n", df.dtypes)

print("\nNull Values:\n", df.isnull().sum())

print("\nDuplicate Rows:", df.duplicated().sum())

Shape: (110527, 14)

Data Types:
 PatientId                     float64
AppointmentID                   int64
Gender                         object
ScheduledDay      datetime64[ns, UTC]
AppointmentDay    datetime64[ns, UTC]
Age                             int64
Neighbourhood                  object
Scholarship                     int64
Hipertension                    int64
Diabetes                        int64
Alcoholism                      int64
Handcap                         int64
SMS_received                    int64
No-show                        object
dtype: object

Null Values:
 PatientId         0
AppointmentID     0
Gender            0
ScheduledDay      0
AppointmentDay    0
Age               0
Neighbourhood     0
Scholarship       0
Hipertension      0
Diabetes          0
Alcoholism        0
Handcap           0
SMS_received      0
No-show           0
dtype: int64

Duplicate Rows: 0


In [10]:
df.rename(columns={
    'Hipertension': 'Hypertension',
    'Handcap': 'Handicap',
    'No-show': 'No_show',
    'ScheduledDay': 'Scheduled_Day',
    'AppointmentDay': 'Appointment_Day'
}, inplace=True)

# Make all column names lowercase
df.columns = df.columns.str.lower()

print(df.columns)

Index(['patientid', 'appointmentid', 'gender', 'scheduled_day',
       'appointment_day', 'age', 'neighbourhood', 'scholarship',
       'hypertension', 'diabetes', 'alcoholism', 'handicap', 'sms_received',
       'no_show'],
      dtype='object')


In [11]:
df['days_between'] = (df['appointment_day'] - df['scheduled_day']).dt.days

In [12]:
df.head()

,patientid,appointmentid,gender,scheduled_day,appointment_day,age,neighbourhood,scholarship,hypertension,diabetes,alcoholism,handicap,sms_received,no_show,days_between
0,2.987250e+13,5642903,F,2016-04-29 18:38:08+00:00,2016-04-29 00:00:00+00:00,62,JARDIM DA PENHA,0,1,0,0,0,0,No,-1
1,5.589978e+14,5642503,M,2016-04-29 16:08:27+00:00,2016-04-29 00:00:00+00:00,56,JARDIM DA PENHA,0,0,0,0,0,0,No,-1
2,4.262962e+12,5642549,F,2016-04-29 16:19:04+00:00,2016-04-29 00:00:00+00:00,62,MATA DA PRAIA,0,0,0,0,0,0,No,-1
3,8.679512e+11,5642828,F,2016-04-29 17:29:31+00:00,2016-04-29 00:00:00+00:00,8,PONTAL DE CAMBURI,0,0,0,0,0,0,No,-1
4,8.841186e+12,5642494,F,2016-04-29 16:07:23+00:00,2016-04-29 00:00:00+00:00,56,JARDIM DA PENHA,0,1,1,0,0,0,No,-1


In [13]:
# Check how many negative values
print("Negative days:", (df['days_between'] < 0).sum())

# Remove them
df = df[df['days_between'] >= 0]

Negative days: 38568


In [14]:
print(df.shape)
print(df.head())

(71959, 15)
       patientid  appointmentid gender             scheduled_day  \
5   9.598513e+13        5626772      F 2016-04-27 08:36:51+00:00   
6   7.336882e+14        5630279      F 2016-04-27 15:05:12+00:00   
7   3.449833e+12        5630575      F 2016-04-27 15:39:58+00:00   
9   7.812456e+13        5629123      F 2016-04-27 12:48:25+00:00   
10  7.345362e+14        5630213      F 2016-04-27 14:58:11+00:00   

             appointment_day  age   neighbourhood  scholarship  hypertension  \
5  2016-04-29 00:00:00+00:00   76       REPÚBLICA            0             1   
6  2016-04-29 00:00:00+00:00   23      GOIABEIRAS            0             0   
7  2016-04-29 00:00:00+00:00   39      GOIABEIRAS            0             0   
9  2016-04-29 00:00:00+00:00   19       CONQUISTA            0             0   
10 2016-04-29 00:00:00+00:00   30  NOVA PALESTINA            0             0   

    diabetes  alcoholism  handicap  sms_received no_show  days_between  
5          0           0 

In [15]:
df


,patientid,appointmentid,gender,scheduled_day,appointment_day,age,neighbourhood,scholarship,hypertension,diabetes,alcoholism,handicap,sms_received,no_show,days_between
5,9.598513e+13,5626772,F,2016-04-27 08:36:51+00:00,2016-04-29 00:00:00+00:00,76,REPÚBLICA,0,1,0,0,0,0,No,1
6,7.336882e+14,5630279,F,2016-04-27 15:05:12+00:00,2016-04-29 00:00:00+00:00,23,GOIABEIRAS,0,0,0,0,0,0,Yes,1
7,3.449833e+12,5630575,F,2016-04-27 15:39:58+00:00,2016-04-29 00:00:00+00:00,39,GOIABEIRAS,0,0,0,0,0,0,Yes,1
9,7.812456e+13,5629123,F,2016-04-27 12:48:25+00:00,2016-04-29 00:00:00+00:00,19,CONQUISTA,0,0,0,0,0,0,No,1
10,7.345362e+14,5630213,F,2016-04-27 14:58:11+00:00,2016-04-29 00:00:00+00:00,30,NOVA PALESTINA,0,0,0,0,0,0,No,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
110522,2.572134e+12,5651768,F,2016-05-03 09:15:35+00:00,2016-06-07 00:00:00+00:00,56,MARIA ORTIZ,0,0,0,0,0,1,No,34
110523,3.596266e+12,5650093,F,2016-05-03 07:27:33+00:00,2016-06-07 00:00:00+00:00,51,MARIA ORTIZ,0,0,0,0,0,1,No,34
110524,1.557663e+13,5630692,F,2016-04-27 16:03:52+00:00,2016-06-07 00:00:00+00:00,21,MARIA ORTIZ,0,0,0,0,0,1,No,40
110525,9.213493e+13,5630323,F,2016-04-27 15:09:23+00:00,2016-06-07 00:00:00+00:00,38,MARIA ORTIZ,0,0,0,0,0,1,No,40


In [16]:
df['patientid'] = df['patientid'].astype(str)
df['appointmentid'] = df['appointmentid'].astype(str)

df.to_csv("cleaned_appointments.csv", index=False)

C:\Users\siddh\AppData\Local\Temp\ipykernel_7120\1264588084.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['patientid'] = df['patientid'].astype(str)
C:\Users\siddh\AppData\Local\Temp\ipykernel_7120\1264588084.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['appointmentid'] = df['appointmentid'].astype(str)


In [17]:
df

,patientid,appointmentid,gender,scheduled_day,appointment_day,age,neighbourhood,scholarship,hypertension,diabetes,alcoholism,handicap,sms_received,no_show,days_between
5,95985133231274.0,5626772,F,2016-04-27 08:36:51+00:00,2016-04-29 00:00:00+00:00,76,REPÚBLICA,0,1,0,0,0,0,No,1
6,733688164476661.0,5630279,F,2016-04-27 15:05:12+00:00,2016-04-29 00:00:00+00:00,23,GOIABEIRAS,0,0,0,0,0,0,Yes,1
7,3449833394123.0,5630575,F,2016-04-27 15:39:58+00:00,2016-04-29 00:00:00+00:00,39,GOIABEIRAS,0,0,0,0,0,0,Yes,1
9,78124564369297.0,5629123,F,2016-04-27 12:48:25+00:00,2016-04-29 00:00:00+00:00,19,CONQUISTA,0,0,0,0,0,0,No,1
10,734536231958495.0,5630213,F,2016-04-27 14:58:11+00:00,2016-04-29 00:00:00+00:00,30,NOVA PALESTINA,0,0,0,0,0,0,No,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
110522,2572134369293.0,5651768,F,2016-05-03 09:15:35+00:00,2016-06-07 00:00:00+00:00,56,MARIA ORTIZ,0,0,0,0,0,1,No,34
110523,3596266328735.0,5650093,F,2016-05-03 07:27:33+00:00,2016-06-07 00:00:00+00:00,51,MARIA ORTIZ,0,0,0,0,0,1,No,34
110524,15576631729893.0,5630692,F,2016-04-27 16:03:52+00:00,2016-06-07 00:00:00+00:00,21,MARIA ORTIZ,0,0,0,0,0,1,No,40
110525,92134931435557.0,5630323,F,2016-04-27 15:09:23+00:00,2016-06-07 00:00:00+00:00,38,MARIA ORTIZ,0,0,0,0,0,1,No,40


In [18]:

df.shape          # how many rows and columns
df.head()         # first 5 rows
df.dtypes         # column data types



patientid                       object
appointmentid                   object
gender                          object
scheduled_day      datetime64[ns, UTC]
appointment_day    datetime64[ns, UTC]
age                              int64
neighbourhood                   object
scholarship                      int64
hypertension                     int64
diabetes                         int64
alcoholism                       int64
handicap                         int64
sms_received                     int64
no_show                         object
days_between                     int64
dtype: object

In [19]:
df.columns.tolist() 

['patientid',
 'appointmentid',
 'gender',
 'scheduled_day',
 'appointment_day',
 'age',
 'neighbourhood',
 'scholarship',
 'hypertension',
 'diabetes',
 'alcoholism',
 'handicap',
 'sms_received',
 'no_show',
 'days_between']

In [20]:
df['no_show'] = df['no_show'].map({'Yes': 1, 'No': 0})

# Verify
df['no_show'].value_counts()
df['no_show'].value_counts(normalize=True).round(3)  

C:\Users\siddh\AppData\Local\Temp\ipykernel_7120\4162201299.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['no_show'] = df['no_show'].map({'Yes': 1, 'No': 0})


no_show
0    0.715
1    0.285
Name: proportion, dtype: float64

In [21]:
print(df['age'].min(), df['age'].max())
print(f"Negative ages: {(df['age'] < 0).sum()}")
print(f"Unrealistic ages: {(df['age'] > 115).sum()}")

# Remove rows with age below 0
df = df[df['age'] >= 0]

0 115
Negative ages: 0
Unrealistic ages: 0


In [31]:
print(df['gender'].value_counts())
print(df['handicap'].value_counts())   # should be 0-4, check for anything weird
print(df['scholarship'].value_counts()) # should only be 0 and 1
print(df['hypertension'].value_counts())
print(df['diabetes'].value_counts())
print(df['alcoholism'].value_counts())
print(df['sms_received'].value_counts())

gender
F    48070
M    23889
Name: count, dtype: int64
handicap
0    70652
1     1307
Name: count, dtype: int64
scholarship
0    65288
1     6671
Name: count, dtype: int64
hypertension
0    56927
1    15032
Name: count, dtype: int64
diabetes
0    66582
1     5377
Name: count, dtype: int64
alcoholism
0    70137
1     1822
Name: count, dtype: int64
sms_received
0    36477
1    35482
Name: count, dtype: int64


In [23]:
# convert to binary
df['handicap'] = df['handicap'].apply(lambda x: 1 if x > 0 else 0)

In [24]:
df['handicap'].value_counts()   # should be 0-4, check for anything weird

handicap
0    70652
1     1307
Name: count, dtype: int64

In [25]:
df['appointment_weekday']     = df['appointment_day'].dt.day_name()
df['appointment_month']       = df['appointment_day'].dt.month_name()
df['appointment_weekday_num'] = df['appointment_day'].dt.dayofweek  # 0=Monday
df['scheduled_hour']          = df['scheduled_day'].dt.hour

In [26]:
bins   = [-1, 17, 35, 60, 115]
labels = ['0-17', '18-35', '36-60', '60+']

df['age_group'] = pd.cut(df['age'], bins=bins, labels=labels)

df['age_group'].value_counts()

age_group
36-60    25380
18-35    16568
0-17     16166
60+      13845
Name: count, dtype: int64

In [27]:
print("Shape after cleaning:", df.shape)
print("Nulls remaining:\n", df.isnull().sum())
print("No-show rate:", df['no_show'].mean().round(3))
print("Date range:", df['appointment_day'].min(), "to", df['appointment_day'].max())

df.dtypes

Shape after cleaning: (71959, 20)
Nulls remaining:
 patientid                  0
appointmentid              0
gender                     0
scheduled_day              0
appointment_day            0
age                        0
neighbourhood              0
scholarship                0
hypertension               0
diabetes                   0
alcoholism                 0
handicap                   0
sms_received               0
no_show                    0
days_between               0
appointment_weekday        0
appointment_month          0
appointment_weekday_num    0
scheduled_hour             0
age_group                  0
dtype: int64
No-show rate: 0.285
Date range: 2016-04-29 00:00:00+00:00 to 2016-06-08 00:00:00+00:00


patientid                               object
appointmentid                           object
gender                                  object
scheduled_day              datetime64[ns, UTC]
appointment_day            datetime64[ns, UTC]
age                                      int64
neighbourhood                           object
scholarship                              int64
hypertension                             int64
diabetes                                 int64
alcoholism                               int64
handicap                                 int64
sms_received                             int64
no_show                                  int64
days_between                             int64
appointment_weekday                     object
appointment_month                       object
appointment_weekday_num                  int32
scheduled_hour                           int32
age_group                             category
dtype: object

In [28]:
df.head()

,patientid,appointmentid,gender,scheduled_day,appointment_day,age,neighbourhood,scholarship,hypertension,diabetes,alcoholism,handicap,sms_received,no_show,days_between,appointment_weekday,appointment_month,appointment_weekday_num,scheduled_hour,age_group
5,95985133231274.0,5626772,F,2016-04-27 08:36:51+00:00,2016-04-29 00:00:00+00:00,76,REPÚBLICA,0,1,0,0,0,0,0,1,Friday,April,4,8,60+
6,733688164476661.0,5630279,F,2016-04-27 15:05:12+00:00,2016-04-29 00:00:00+00:00,23,GOIABEIRAS,0,0,0,0,0,0,1,1,Friday,April,4,15,18-35
7,3449833394123.0,5630575,F,2016-04-27 15:39:58+00:00,2016-04-29 00:00:00+00:00,39,GOIABEIRAS,0,0,0,0,0,0,1,1,Friday,April,4,15,36-60
9,78124564369297.0,5629123,F,2016-04-27 12:48:25+00:00,2016-04-29 00:00:00+00:00,19,CONQUISTA,0,0,0,0,0,0,0,1,Friday,April,4,12,18-35
10,734536231958495.0,5630213,F,2016-04-27 14:58:11+00:00,2016-04-29 00:00:00+00:00,30,NOVA PALESTINA,0,0,0,0,0,0,0,1,Friday,April,4,14,18-35


In [29]:
df['patientid'] = df['patientid'].astype(str).str.replace('.0', '', regex=False)
df['appointmentid'] = df['appointmentid'].astype(str).str.replace('.0', '', regex=False)

In [30]:
df.to_csv('cleaned_appointments.csv', index=False)
print(f"Saved {df.shape[0]} rows, {df.shape[1]} columns")

Saved 71959 rows, 20 columns



## Data Cleaning Summary
- Started with 110,527 rows, 14 columns
- Removed 385688 rows with negative days_between (scheduling errors) 
- Converted no_show to binary: 1 = no-show, 0 = showed up
- Fixed column names (corrected misspellings: Hipertension, Handcap)
- Engineered: days_between, age_group, appointment_weekday, appointment_month
- Final dataset: [71959] rows, [20] columns
- Overall no-show rate: ~20.2%